RestDocs를 적용하기 위해 gradle을 설정한다.

스토리 포인트: 1일

gradle 버전을 7.0 버전으로 올려보고 Restdocs 설정을 해보려 했으나 vaadin 관련 호환 이슈때문에 6.5.1 버전으로 진행해보려 합니다.

**장바구니 미션에서 사용한 코드를 이용해 gradle 버전을 7.0으로 올려보고 `org.assciidoctor.convert` 를 `org.asciidoctor.jvm.convert` 로 변경해서 진행한 결과 문제가 없음을 확인했습니다.**

지원플랫폼 코드에서는 6.5.1 버전으로 작업을 진행하려고 하는데 @woowahan-pjs 는 어떻게 생각하시나요?

### gradle 6.5.1 에서 7.0으로 변경 후 빌드시 발생 오류
``` bash
$ ./gradlew build --warning-mode all

> Configure project :
You are using one or more deprecated Asciidoctor Gradle plugin features.
These will be removed in 3.0 of these plugins.
To help you migrate we have compiled some tips for you based upon your current usage.
- org.asciidoctor.convert:
  - 'org.asciidoctor.convert' is deprecated. When you have time please switch over to 'org.asciidoctor.jvm.convert'.
  - jcenter() is no longer added by default. If you relied on this behaviour in the past, please add jcenter() to the repositories block.


The RepositoryHandler.jcenter() method has been deprecated. This is scheduled to be removed in Gradle 8.0. JFrog announced JCenter's shutdown in February 2021. Use mavenCentral() instead. Consult the upgrading guide for further information: https://docs.gradle.org/7.0/userguide/upgrading_version_6.html#jcenter_deprecation
Could not resolve compiler classpath. Check if Kotlin Gradle plugin repository is configured in root project 'apply'.

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':compileKotlin'.
> Failed to notify dependency resolution listener.
   > Could not resolve all dependencies for configuration ':detachedConfiguration3'.
      > Using insecure protocols with repositories, without explicit opt-in, is unsupported. Switch Maven repository 'maven(http://maven.vaadin.com/vaadin-addons)' to redirect to a secure protocol (like HTTPS) or allow insecure protocols. See https://docs.gradle.org/7.0/dsl/org.gradle.api.artifacts.repositories.UrlArtifactRepository.html#org.gradle.api.artifacts.repositories.UrlArtifactRepository:allowInsecureProtocol for more details. 
   > Could not resolve all dependencies for configuration ':detachedConfiguration4'.
      > Using insecure protocols with repositories, without explicit opt-in, is unsupported. Switch Maven repository 'maven(http://maven.vaadin.com/vaadin-addons)' to redirect to a secure protocol (like HTTPS) or allow insecure protocols. See https://docs.gradle.org/7.0/dsl/org.gradle.api.artifacts.repositories.UrlArtifactRepository.html#org.gradle.api.artifacts.repositories.UrlArtifactRepository:allowInsecureProtocol for more details. 

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 554ms
```

--- a/.gitignore
+++ b/.gitignore
@@ -42,3 +42,6 @@ pnpmfile.js
 docker/db/mysql/data/
 docker/db/mysql/init/*
 !docker/db/mysql/init/init.sql
+
+### Spring REST Docs ###
+**/src/main/resources/static/docs/

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -9,7 +9,7 @@ plugins {
     kotlin("plugin.jpa") version kotlinVersion
     id("org.jlleitschuh.gradle.ktlint") version "10.1.0"
     id("com.vaadin") version "0.8.0"
-    id("org.asciidoctor.convert") version "2.4.0"
+    id("org.asciidoctor.jvm.convert") version "3.3.2"
     id("org.flywaydb.flyway") version "7.12.0"
 }
 
@@ -23,6 +23,8 @@ repositories {
         url = uri("http://maven.vaadin.com/vaadin-addons")
     }
 }
+
+val asciidoctorExt by configurations.creating
 
 extra["vaadinVersion"] = "14.3.3"
 
@@ -51,7 +53,7 @@ dependencies {
         exclude(group = "org.mockito")
     }
     testImplementation("com.ninja-squad:springmockk:2.0.3")
-    asciidoctor("org.springframework.restdocs:spring-restdocs-asciidoctor")
+    asciidoctorExt("org.springframework.restdocs:spring-restdocs-asciidoctor")
     testImplementation("org.springframework.restdocs:spring-restdocs-mockmvc")
 }
 
@@ -89,13 +91,16 @@ tasks {
         outputs.dir(snippetsDir)
     }
     asciidoctor {
-        inputs.dir(snippetsDir)
         dependsOn(test)
+        configurations("asciidoctorExt")
+        inputs.dir(snippetsDir)
+    }
+    register<Copy>("copyDocument") {
+        dependsOn(asciidoctor)
+        from(file("build/docs/asciidoc/index.html"))
+        into(file("src/main/resources/static/docs"))
     }
     bootJar {
-        dependsOn(asciidoctor)
-        from("$snippetsDir/html5") {
-            into("static/docs")
-        }
+        dependsOn("copyDocument")
     }
 }

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -24,9 +24,10 @@ repositories {
     }
 }
 
-val asciidoctorExt by configurations.creating
-
 extra["vaadinVersion"] = "14.3.3"
+
+val asciidoctorExt: Configuration by configurations.creating
+val snippetsDir by extra { "build/generated-snippets" }
 
 dependencies {
     implementation("org.springframework.boot:spring-boot-starter-data-jpa")
@@ -63,10 +64,6 @@ dependencyManagement {
     }
 }
 
-val snippetsDir by extra {
-    file("build/generated-snippets")
-}
-
 tasks {
     withType<Test> {
         useJUnitPlatform()
@@ -91,16 +88,17 @@ tasks {
         outputs.dir(snippetsDir)
     }
     asciidoctor {
-        dependsOn(test)
-        configurations("asciidoctorExt")
         inputs.dir(snippetsDir)
+        configurations("asciidoctorExt")
+        dependsOn(test)
+        baseDirFollowsSourceFile()
     }
-    register<Copy>("copyDocument") {
+    val copyDocs = register<Copy>("copyDocs") {
         dependsOn(asciidoctor)
-        from(file("build/docs/asciidoc/index.html"))
-        into(file("src/main/resources/static/docs"))
+        from("${asciidoctor.get().outputDir}/index.html")
+        into("src/main/resources/static/docs")
     }
     bootJar {
-        dependsOn("copyDocument")
+        dependsOn(copyDocs)
     }
 }

--- a/src/docs/asciidoc/index.adoc
+++ b/src/docs/asciidoc/index.adoc
@@ -3,3 +3,5 @@
 :toc: left
 :toclevels: 1
 :sectlinks:
+
+== Service-Apply

--- a/src/docs/asciidoc/index.adoc
+++ b/src/docs/asciidoc/index.adoc
@@ -1,0 +1,5 @@
+:doctype: book
+:source-highlighter: highlightjs
+:toc: left
+:toclevels: 1
+:sectlinks:

